File: 888.fair-candy-swap.cpp
/*
 * @lc app=leetcode id=888 lang=cpp
 *
 * {888} Fair Candy Swap
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<int> fairCandySwap(vector<int>& A, vector<int>& B) {
    int totalA = 0, totalB = 0;
    for (int size : A) {
      totalA += size;
    }
    for (int size : B) {
      totalB += size;
    }
    int diff = (totalA - totalB);
    for (int sizeA : A) {
      for (int sizeB : B) {
        if ((sizeA - sizeB) * 2 == diff) {
          return {sizeA, sizeB};
        }
      }
    }
    return {-1, -1};
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> A = {1, 1}, B = {2, 2};
  vector<int> exp = {1, 2};
  ASSERT_EQ(s.fairCandySwap(A, B), exp);

  // Example 2:
  A = {1, 2}, B = {2, 3};
  exp = {1, 2};
  ASSERT_EQ(s.fairCandySwap(A, B), exp);

  // Example 3:
  A = {2}, B = {1, 3};
  exp = {2, 3};
  ASSERT_EQ(s.fairCandySwap(A, B), exp);

  // Example 4:
  A = {1, 2, 5}, B = {2, 4};
  exp = {5, 4};
  ASSERT_EQ(s.fairCandySwap(A, B), exp);

  return 0;
}

File: 888.fair-candy-swap.rs
/*
 * @lc app=leetcode id=888 lang=rust
 *
 * [888] Fair Candy Swap
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn fair_candy_swap(alice_sizes: Vec<i32>, bob_sizes: Vec<i32>) -> Vec<i32> {
        let sum_a: i32 = alice_sizes.iter().sum();
        let sum_b: i32 = bob_sizes.iter().sum();
        let diff = (sum_a - sum_b) / 2;
        let set_b: std::collections::HashSet<i32> = bob_sizes.iter().cloned().collect();
        for alice_size in alice_sizes {
            if set_b.contains(&(alice_size - diff)) {
                return vec![alice_size, alice_size - diff];
            }
        }
        vec![-1, -1]
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::fair_candy_swap(vec![1, 1], vec![2, 2]),
        vec![1, 2]
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::fair_candy_swap(vec![1, 2], vec![2, 3]),
        vec![1, 2]
    );
    assert_eq!(Solution::fair_candy_swap(vec![2], vec![1, 3]), vec![2, 3]);
}
